framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -593,195 +484,306 @@ public void standardizeVariable(AnnotatedTypeMirror type, Element elt) {
-     * @param type the type to parse; is side-effected by this method
+     * <p>For example, if the contract is {@code @EnsuresKeyFor(value = "this.field", map =
+     * "this.map")}, this method applies {@code stringToJava} to "this.map" and returns a new {@code
+     * KeyFor} annotation with the result.
+     *
+     * @param annoFromContract an annotation from a contract
+     * @param stringToJavaExpr function used to convert strings to {@link JavaExpression}s
+     * @param errorTree if non-null, where to report any errors that occur when parsing the
+     *     dependent type annotation; if null, report no errors
+     * @return the viewpoint-adapted annotation, or the argument if it is not a dependant type
+     *     annotation
      */
-    private void parseToPath(TreePath localVarPath, AnnotatedTypeMirror type) {
-        Tree enclosingClass = TreePathUtil.enclosingClass(localVarPath);
-        TypeMirror enclosingType = TreeUtils.typeOf(enclosingClass);
-
-        JavaExpression receiver = JavaExpression.getPseudoReceiver(localVarPath, enclosingType);
-
-        JavaExpressionContext localContext =
-                new JavaExpressionContext(
-                        receiver,
-                        JavaExpression.getParametersOfEnclosingMethod(localVarPath),
-                        factory.getChecker());
-        standardizeAtm(localContext, localVarPath, type, /*removeErroneousExpressions=*/ false);
+    public AnnotationMirror viewpointAdaptQualifierFromContract(
+            AnnotationMirror annoFromContract,
+            StringToJavaExpression stringToJavaExpr,
+            @Nullable Tree errorTree) {
+        if (!hasDependentAnnotations()) {
+            return annoFromContract;
+        }
+
+        AnnotationMirror standardized = map(stringToJavaExpr, annoFromContract);
+        if (standardized == null) {
+            return annoFromContract;
+        }
+        if (errorTree != null) {
+            checkAnnotation(standardized, errorTree);
+        }
+        return standardized;
     }
 
     /**
-     * Viewpont-adapt the dependent types in {@code type} using the {@code context} provided.
+     * Viewpoint-adapt all dependent type annotations to the method declaration, {@code
+     * methodDeclTree}. This changes occurrences of formal parameter names to the "#2" syntax, and
+     * it removes expressions that contain other local variables.
      *
-     * @param context the context
-     * @param type the type to viewpoint-adapt; is side-effected by this method
+     * @param methodDeclTree the method declaration to which the annotations are viewpoint-adapted
+     * @param atm type to viewpoint-adapt; is side-effected by this method
      */
-    private void viewpointAdaptToContext(JavaExpressionContext context, AnnotatedTypeMirror type) {
-        standardizeAtm(context, null, type, /*removeErroneousExpressions=*/ false);
+    public void delocalize(MethodTree methodDeclTree, AnnotatedTypeMirror atm) {
```

Comment:
What is the relationship to `atMethodDeclaration`?  Why isn't this method named `atMethodDeclaration` too?